python/pfs/drp/qa/dmResiduals.py

```diff
@@ -301,12 +320,95 @@ def to_dict(self):
         )
 
 
+def getGoodLines(
+    lines: ArcLineSet, dispersion: float | None, adjustDMConfig: Config, log: Logger | None = None
+) -> np.ndarray:
+    """Get the good lines.
+
+    Parameters
+    ----------
+
+    lines : `ArcLineSet`
+        The arc lines.
+    dispersion : `float`, optional
+        The dispersion. Default is None.
+    adjustDMConfig : `Config`
+        Configuration used for the detector map adjustment.
+    log : `Logger`, optional
+        The logger for the class object. Default is None.
+
+    Returns
+    -------
+    good : `np.ndarray`
+        The index of the good lines.
+    """
+    log.debug(f"Scrubbing data using config={adjustDMConfig.toDict()}")
+    isTrace = lines.description == "Trace"
+    isLine = ~isTrace
+    numTraceLines = len(set(lines[isTrace].fiberId))
+    numArcLines = len(set(lines[isLine].fiberId))
+
+    log.debug(f"{isTrace.sum()} line centroids for {numTraceLines} traces")
+    log.debug(f"{isLine.sum() + isTrace.sum()} lines in list")
+
+    def getCounts():
+        """Provide a list of counts of different species"""
+        return getDescriptionCounts(lines.description, good)
+
+    good = lines.flag == 0
+    log.debug(f"{good.sum()} good lines after initial flags ({getCounts()})")
+
+    if isLine.sum() > 0:
+        log.info("Found lamp species, ignoring traces.")
+        good &= isLine
+        log.debug(f"{good.sum()} good lines after ignoring traces ({getCounts()})")
+
+    good &= (lines.status & ReferenceLineStatus.fromNames(*adjustDMConfig.lineFlags)) == 0
+    log.debug(f"{good.sum()} good lines after line flags ({getCounts()})")
+
+    good &= np.isfinite(lines.x) & np.isfinite(lines.y)
+    good &= np.isfinite(lines.xErr) & np.isfinite(lines.yErr)
+
+    if hasattr(lines, "slope"):
+        good &= np.isfinite(lines.slope) | ~isTrace
+    log.debug(f"{good.sum()} good lines after finite positions ({getCounts()})")
+
+    if adjustDMConfig.minSignalToNoise > 0:
+        good &= np.isfinite(lines.flux) & np.isfinite(lines.fluxErr)
+        log.debug(f"{good.sum()} good lines after finite intensities ({getCounts()})")
+
+        with np.errstate(invalid="ignore", divide="ignore"):
+            sn = lines.flux / lines.fluxErr
+            mean_sn = np.nanmean(sn[good])
+            log.debug(f"Using mean SN={mean_sn:.02f} instead of config {adjustDMConfig.minSignalToNoise}")
+            good &= sn > mean_sn
```

Comment:
@kiyoyabe here is the SN filtering that is applied, which is just a simple mean.  The logs for this section are:

```
(dmResiduals:{instrument: 'PFS', arm: 'b', spectrograph: 3, visit: 115835, ...})(dmResiduals.py:364) - 16429 good lines after ignoring traces (KrI: 16429)
(dmResiduals:{instrument: 'PFS', arm: 'b', spectrograph: 3, visit: 115835, ...})(dmResiduals.py:367) - 15279 good lines after line flags (KrI: 15279)
(dmResiduals:{instrument: 'PFS', arm: 'b', spectrograph: 3, visit: 115835, ...})(dmResiduals.py:374) - 15279 good lines after finite positions (KrI: 15279)
(dmResiduals:{instrument: 'PFS', arm: 'b', spectrograph: 3, visit: 115835, ...})(dmResiduals.py:378) - 15094 good lines after finite intensities (KrI: 15094)
(dmResiduals:{instrument: 'PFS', arm: 'b', spectrograph: 3, visit: 115835, ...})(dmResiduals.py:383) - Using mean SN=10.10 instead of config 20.0
(dmResiduals:{instrument: 'PFS', arm: 'b', spectrograph: 3, visit: 115835, ...})(dmResiduals.py:386) - 7713 good lines after min SN=10.10 (KrI: 7713)
(dmResiduals:{instrument: 'PFS', arm: 'b', spectrograph: 3, visit: 115835, ...})(dmResiduals.py:392) - 4952 good lines after maxCentroidError=0.15 centroid errors (KrI: 4952)
```